Reimplement MSW wxPopupWindow

vadz opened the pull request:
This allows having functional controls inside popups.

I don't have time to do this now, but I'd be curious if someone could try setting `TRANSIENT_POPUPWIN_IS_PERFECT` to 1 in `src/common/combocmn.cpp` for MSW now -- it might actually have a chance to work after these changes.

vadz committed:
Reimplement wxPopupWindow as a WS_POPUP window under MSW

Don't use the child window of the desktop window for popup windows under
MSW, while this worked in simplest cases, it didn't allow having
functional controls inside a wxPopupWindow as e.g. wxTextCtrl didn't
accept input it at all if created as a child of such window.

Instead, switch to using a top-level window, with WS_POPUP style, and
fix the problem with the loss of activation by explicitly pretending to
still be active in the owner window when losing activation to our own
popup (thanks to Barmak Shemirani for providing this solution).

Also use an MSW-specific and much simpler implementation of detecting
when the popup should be dismissed in wxPopupTransientWindow: instead of
capturing mouse or tracking focus, just react to activation loss
directly.

Add a wxTextCtrl to the popup in samples/popup to show that editing it
works now.

vadz committed:
Don't assert in focus code for buttons inside a wxPopupWindow

IsTopLevel() returns true for wxPopupWindow, even if it's not a subclass
of wxTopLevelWindow, so GetTLWParentIfNotBeingDeleted() asserted when
called with a button inside a wxPopupWindow.

Just return null from it instead for now. A better solution could be to
return wxNonOwnedWindow from GetTLWParentIfNotBeingDeleted() (which
would need to be renamed to something more suitable) and move the
{Get,Set}TmpDefaultItem() methods into it.

vadz commented on include/wx/popupwin.h:
```diff
@@ -104,32 +96,82 @@ class WXDLLIMPEXP_CORE wxPopupTransientWindow : public wxPopupWindow
+    virtual void Dismiss() wxOVERRIDE;
+
+    // Overridden to grab the input on some plaforms
+    virtual bool Show( bool show = true ) wxOVERRIDE;
```
This line was moved, not added, and I prefer to avoid changing formatting in such cases.

vadz commented on src/msw/button.cpp:
```diff
@@ -252,28 +252,14 @@ wxWindow *wxButton::SetDefault()
 
-    return tlw;
+    // Note that this may still return null for a button inside wxPopupWindow.
+    return wxDynamicCast(win, wxTopLevelWindow);
```
Oops, yes, you're right, thanks for noticing this!

kivadiu commented:
It took me a while to check your PR because although it became evident that it fixed the issue that prevented text to be entered in the text control of the popup window, at the same time, I could not see the popup window in my own code where I use to click on a base control to trigger the popop.
The fact is that I called `Popup(this)` (`focus == this`) which used to work and do not work anymore.
Nevertheless, I removed this and call `Popup()` which removes the focus from the base control (I preferred the old behaviour but that's not critical).

One thing does not work however: if my base control is close to the bottom of the frame and the text control at the bottom of the popup (with some other control at the top), when I click in the text control of the popup, the scrolled window where the base control is in scrolls down as if I had clicked in the scrolled window and not in the text control of the popup window... I will try to make a sample to show that.

vadz committed:
Block child window events propagation beyond wxPopupWindow

Set wxWS_EX_BLOCK_EVENTS to prevent events bubbling up to the popup
parent as it doesn't make much sense.

See #18243.

MaartenBent commented:
This introduces the following new warning:
```
wxWidgets\src\msw\popupwin.cpp:36:23: warning: 'wxCurrentPopupWindow' initialized and declared 'extern'
 extern wxPopupWindow* wxCurrentPopupWindow = NULL;
                       ^~~~~~~~~~~~~~~~~~~~
```
`Extern` is only needed in the header.

vadz commented:
@MaartenBent Fixed in a316942c18e516ba212dd7f537173485285c7a90, thanks.
